haystack/utils/reflection.py

```diff
@@ -33,3 +37,47 @@ def wrapper_invocation_counter(*args, **kwargs):
 
     wrapper_invocation_counter.counter = 0
     return wrapper_invocation_counter
+
+
+def retry_with_exponential_backoff(
+    initial_delay: float = 1,
+    exponential_base: float = 2,
+    jitter: bool = True,
+    max_retries: int = 10,
+    errors: tuple = (OpenAIRateLimitError,),
+):
+    """Retry a function with exponential backoff."""
+
+    def decorator(function):
+        def wrapper(*args, **kwargs):
+            # Initialize variables
+            num_retries = 0
+            delay = initial_delay
+
+            # Loop until a successful response or max_retries is hit or an exception is raised
+            while True:
+                try:
+                    return function(*args, **kwargs)
+
+                # Retry on specified errors
+                except errors as e:
+                    # Increment retries
+                    num_retries += 1
+
+                    # Check if max retries has been reached
+                    if num_retries > max_retries:
+                        raise Exception(f"Maximum number of retries ({max_retries}) exceeded.")
+
+                    # Increment the delay
+                    delay *= exponential_base * (1 + jitter * random())
+
+                    # Sleep for the delay
+                    time.sleep(delay)
```

Comment:
I think it may be helpful to add a logging statement here to let the user know that exponential backoff is being used and how long the current wait time is. What do you think?

Reply:
Yes, good point